### PR TITLE
Implemented the Item Mapper functionality

### PR DESCRIPTION
**Pull request by Raf-atmire**
In this PR we have implemented the functionality to map Items to Collections.
This can be achieved by using the following calls:
- /api/core/items/{itemUuid}/mappedCollections/{collectionUuid} POST call to create a new mapping
- /api/core/items/{itemUuid}/mappedCollections/{collectionUuid} DELETE call to delete a mapping 
- /api/core/items/{itemUuid}/mappedCollections to retrieve all the mappedCollections for this Item
- /api/core/collection/{collectionUuid}/mappedItems to retrieve all the Items mapped to this collection for which the owningCollection is not the given collection

This functionality includes changes to the services, DAOs and controllers to achieve the functionality as well as the creation of tests to ensure that the functionality doesn't break

Jira link: https://jira.duraspace.org/browse/DS-4097
Rest contract link: https://github.com/DSpace/Rest7Contract/pull/52

### CODE REVIEW

**Comment by tdonohue** [comment]
@Raf-atmire & @benbosman  : Note, this PR is now "unblocked" but it requires updating per the *new* REST Contract approved in https://github.com/DSpace/Rest7Contract/pull/52

**Comment by KevinVdV** [comment]
This PR is now ready for review.

**Review by benbosman — APPROVED** [review]
With the latest changes in place, this looks good to me
The functionality matches the REST Contract
The ITs are present and cover all the use cases which seem relevant to me
The updates to the ItemService are also tested in the ItemTest (apart from the indirect tests in the REST ITs)

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingCollectionRestController.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingCollectionRestController.java]: Please add JavaDocs to this new class & all public methods.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java]: Please add Javadocs to this new class and all public methods

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java]: It seems odd to me that this endpoint is `/collections/[uuid]/mappingItems`, while the Item endpoint is `/items/[uuid]/mappedCollections`.  Shouldn't these be named similarly?

I'd recommend renaming this to be `/collections/[uuid]/mappedItems`.  I realize that may also require minor changes to the REST Contract, but it seems more consistent

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingCollectionResourceWrapperHalLinkFactory.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingCollectionResourceWrapperHalLinkFactory.java]: Please add Javadocs to this new class

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingCollectionRestHalLinkFactory.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingCollectionRestHalLinkFactory.java]: Please add Javadocs to this new class

File: dspace-api/src/main/java/org/dspace/content/dao/ItemDAO.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/dao/ItemDAO.java]: Please add JavaDocs for these two new public methods.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingItemResourceWrapperHalLinkFactory.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingItemResourceWrapperHalLinkFactory.java]: Please add Javadocs to this new class & all public methods

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingItemRestHalLinkFactory.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingItemRestHalLinkFactory.java]: Please add Javadocs to this new class

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingCollectionRestWrapper.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingCollectionRestWrapper.java]: Please add Javadocs to this new class & all public methods

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingItemRestWrapper.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingItemRestWrapper.java]: Please add Javadocs to this new class

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/hateoas/MappingCollectionResourceWrapper.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/hateoas/MappingCollectionResourceWrapper.java]: Please add Javadocs to this new class

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/hateoas/MappingItemResourceWrapper.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/hateoas/MappingItemResourceWrapper.java]: Please add Javadocs to this new class

**Review by tdonohue — COMMENTED** [review]
I've given this a quick review and found that most of the code is currently missing Javadocs. We'll need to get those added in before we can merge.  A few other minor comments below inline.

**Comment by tdonohue** [comment]
@Raf-atmire : this seems to have a minor conflict from recent merges today.  If you can get the conflict resolved, I'll do another review tomorrow (of the newly added javadocs) and see if we can get this merged.  Thanks!

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java
Comment (Raf-atmire) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java]: Hi @tdonohue thank you for the feedback. That would indeed make a lot more sense and I'll change this in the following commit. However I had an additional question as it may make sense to rename everything from 'mappingItems' to 'mappedItems' then if we're changing the endpoint syntax. For reference: https://github.com/DSpace/DSpace/pull/2282/files#diff-c56be5b7c98dee81bed835af5bf7cb6dR45
Currently they get embedded under the 'mappingItems' title, should we change this to 'mappedItems' as well along with refactoring class names etc? Just to make it more consistent with the proposed change as that would make everything even more consistent

Comment (tdonohue) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java]: @Raf-atmire : Yes, I agree and I'd be in favor of renaming all "mappingItems" to "mappedItems" for consistency.

**Comment by KevinVdV** [comment]
@tdonohue The issues have been resolved, this PR is now ready for testing.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java]: @benbosman  and @Raf-atmire , I've just realized this endpoint isn't described in our REST Contract: https://github.com/DSpace/Rest7Contract/blob/master/collections.md

It's not a big deal for the purposes of approving this PR.  But, we need a separate PR for the REST Contract, please. 

For the REST Contract PR, we'll also need to ensure we make it clear the `/collections/[uuid]/mappedItems` endpoint is READ-ONLY.  Item mappings can only be modified via `/items/[uuid]/mappedCollections`.   I like that approach, we just need to document it.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java]: This comment should say `mappedItems`

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java]: This example should say `mappedItems`

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/hateoas/CollectionResource.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/hateoas/CollectionResource.java]: This should say "mappedItems"

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/hateoas/MappingItemResourceWrapper.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/hateoas/MappingItemResourceWrapper.java]: This should say "mappedItems"

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingCollectionRestController.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingCollectionRestController.java]: Shouldn't this class be named `MappedCollectionRestController`?

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java]: Shouldn't this class be named `MappedItemRestController`?

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingCollectionResourceWrapperHalLinkFactory.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingCollectionResourceWrapperHalLinkFactory.java]: Shouldn't this class be named `MappedCollectionResourceWrapperHalLinkFactory`?

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingCollectionRestHalLinkFactory.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingCollectionRestHalLinkFactory.java]: Shouldn't this class be named `MappedCollectionRestHalLinkFactory`?

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingItemResourceWrapperHalLinkFactory.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingItemResourceWrapperHalLinkFactory.java]: Shouldn't this class be named `MappedItemResourceWrapperHalLinkFactory`?

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingItemRestHalLinkFactory.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingItemRestHalLinkFactory.java]: Shouldn't this class be named `MappedItemRestHalLinkFactory`?

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingCollectionRestWrapper.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingCollectionRestWrapper.java]: Shouldn't this class be named `MappedCollectionRestWrapper`?

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingItemRestWrapper.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingItemRestWrapper.java]: Shouldn't this class be named `MappedItemRestWrapper`?

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingCollectionRestWrapper.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingCollectionRestWrapper.java]: "getMappedCollectionRestList()" / "setMappedCollectionRestList()"?

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingItemRestWrapper.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingItemRestWrapper.java]: Again "Mapped" for getter/setter

**Review by tdonohue — COMMENTED** [review]
@Raf-atmire, @KevinVdV and @benbosman : I hate to get picky here, but we still need to make this PR a bit more consistent with the REST Contract, etc.  It's a bit confusing to me that "Mapping" is used throughout class names, when the endpoints all use "Mapped".   To me, the proper terminology seems like it should be "MappedCollection" and "MappedItem" -- a "MappingCollection" or "MappingItem" isn't a clear name. 

So, I'd like to see more consistency here in the Class names.   That said, the methods in `ItemServiceImpl` can keep the "Mapping" terminology as it comes at the end... so, a "CollectionMapping" makes sense, but a "MappingCollection" is a confusing term.

Overall, honestly, I think this code is very close to merger.  It's well written code and tests.  It's really just the lack of consistent terminology that I'm having trouble with.  So, apologies if this seems really picky.  I'd just like to clean this up now, as I suspect it will be confusing to others as well.

Some terminology notes are inline. Though I may not have caught all of them.  Essentially, I'm just asking for "MappingCollection" to be renamed "MappedCollection" everywhere, and "MappingItem" to be renamed "MappedItem" everywhere.  So, hopefully it's a simple find & replace.

**Comment by tdonohue** [comment]
@Raf-atmire : Just a note that this will need rebasing after the merger of #2459 as the `dspace-spring-rest` folder/webapp has been moved to `dspace-server-webapp`.

**Review by tdonohue — APPROVED** [review]
👍 Looks good to me now.  Thanks @Raf-atmire and @benbosman 

Could someone create a (small) REST Contract PR to document the new `/api/core/collection/{collectionUuid}/mappedItems` Read-only endpoint created in this PR?  That endpoint is undocumented, but I agree that it's needed.  I'm fine with merging this PR, but I'd simply ask we "resync" the REST Contract in the near future.

**Comment by benbosman** [comment]
I've created the additional PR at https://github.com/DSpace/Rest7Contract/pull/70

**Comment by tdonohue** [comment]
Thanks @Raf-atmire and @benbosman.  Looks good now. Merging.

**Pull request merged**
